astropy/modeling/core.py

```diff
@@ -3356,8 +3356,7 @@ def replace_submodel(self, name, model):
                     branch = getattr(branch, node)
                 setattr(branch, tree[-1], model)
                 model = CompoundModel(branch.op, branch.left, branch.right,
```

Comment:
Is it worth emitting a warning when an inverse is lost?  Just to prompt the user to choose between setting a new inverse or deliberately continuing without it.

Incidentally, is it intentional that we're also discarding any user-defined bounding box?

Reply:
I think `bounding_box`es will be propagated up the nodes as the model is rebuilt, right? If there's a user-defined one on a higher node, then it will be removed, just like a user-defined inverse. So this appears consistent.

I'm genuinely not sure about the answer to the first question.

Reply:
In my mind a model and its inverse go together. If one is replacing a model they are probably thinking about the inverse as well. I think there's no need for a warning here. However, we can reconsider in the future if we find this is causing confusion.
That said, it may be good to mention what happens to the inverse in the documentation.
Other than that it looks good to me. I agree it should go in 4.1.